File: plugins/skill-intake/scripts/keychain_get_secret.js
#!/usr/bin/env node
// keychain_get_secret.js
// macOS Keychain から Notion トークンを取得する唯一の経路。
// 他の script / agent はこのモジュール経由でのみトークンに触れること。
// exit codes: 0=OK, 2=INPUT_ERROR, 44=KEYCHAIN_ERROR

'use strict';

const { spawnSync } = require('node:child_process');

const SERVICE = process.env.INTAKE_KEYCHAIN_SERVICE || 'notion-api-key';
const ACCOUNT = process.env.INTAKE_KEYCHAIN_ACCOUNT || 'skill-intake-interviewer';

function getSecret({ service = SERVICE, account = ACCOUNT } = {}) {
  if (process.platform !== 'darwin') {
    const err = new Error(`unsupported platform: ${process.platform} (macOS only)`);
    err.code = 'PLATFORM_UNSUPPORTED';
    err.exit = 44;
    throw err;
  }
  const res = spawnSync('/usr/bin/security', [
    'find-generic-password', '-s', service, '-a', account, '-w',
  ], { encoding: 'utf8' });

  if (res.status !== 0) {
    const err = new Error(
      `Keychain lookup failed (service=${service}, account=${account}): ` +
      (res.stderr || '').trim()
    );
    err.code = 'KEYCHAIN_NOT_FOUND';
    err.exit = 44;
    throw err;
  }
  const token = (res.stdout || '').replace(/\n$/, '');
  if (!token) {
    const err = new Error('Keychain returned empty token');
    err.code = 'KEYCHAIN_EMPTY';
    err.exit = 44;
    throw err;
  }
  return token;
}

function maskToken(t) {
  if (!t) return '(empty)';
  const prefix = t.slice(0, 4);
  return `${prefix}... (len=${t.length})`;
}

module.exports = { getSecret, maskToken, SERVICE, ACCOUNT };

if (require.main === module) {
  const args = new Set(process.argv.slice(2));
  try {
    const t = getSecret();
    if (args.has('--check')) {
      console.log(`OK ${maskToken(t)}`);
    } else if (args.has('--print-unsafe')) {
      process.stdout.write(t);
    } else {
      console.log(`OK service=${SERVICE} account=${ACCOUNT} ${maskToken(t)}`);
      console.log('hint: use --print-unsafe to emit raw token (avoid in shared terminals)');
    }
    process.exit(0);
  } catch (e) {
    console.error(`[keychain_get_secret] ${e.message}`);
    process.exit(e.exit || 1);
  }
}

File: plugins/skill-intake/scripts/keychain_get_secret.py
#!/usr/bin/env python3
"""macOS Keychain から Notion トークンを取得する唯一の経路。"""

import argparse
import os
import subprocess
import sys
from pathlib import Path

DEFAULT_SERVICE = 'notion-api-key'
DEFAULT_ACCOUNT = 'skill-intake'


def _load_config_if_available():
    sys.path.insert(0, str(Path(__file__).resolve().parent))
    import notion_config
    return notion_config.load_config()


def _default_service():
    """毎呼出 env を再評価 (module-level 定数だと同一プロセスでの repo 切替に追随できない)。
    .notion-config.json がある場合は keychain_service を尊重する。
    """
    if os.environ.get('INTAKE_KEYCHAIN_SERVICE'):
        return os.environ['INTAKE_KEYCHAIN_SERVICE']
    cfg = _load_config_if_available()
    if cfg and cfg.get('keychain_service'):
        return cfg['keychain_service']
    return DEFAULT_SERVICE


def _default_account():
    if os.environ.get('INTAKE_KEYCHAIN_ACCOUNT'):
        return os.environ['INTAKE_KEYCHAIN_ACCOUNT']
    cfg = _load_config_if_available()
    if cfg and cfg.get('keychain_account'):
        return cfg['keychain_account']
    return os.environ.get('INTAKE_KEYCHAIN_ACCOUNT', DEFAULT_ACCOUNT)


# 後方互換 alias。実行時解決は _default_* を使う。
SERVICE = DEFAULT_SERVICE
ACCOUNT = DEFAULT_ACCOUNT


class KeychainError(Exception):
    def __init__(self, message, exit_code=44):
        super().__init__(message)
        self.exit_code = exit_code


def get_secret(service=None, account=None):
    service = service or _default_service()
    account = account or _default_account()
    if sys.platform != 'darwin':
        raise KeychainError(f'unsupported platform: {sys.platform} (macOS only)')
    res = subprocess.run(
        ['/usr/bin/security', 'find-generic-password', '-s', service, '-a', account, '-w'],
        capture_output=True, text=True,
    )
    if res.returncode != 0:
        raise KeychainError(
            f'Keychain lookup failed (service={service}, account={account}): '
            + (res.stderr or '').strip()
        )
    token = (res.stdout or '').rstrip('\n')
    if not token:
        raise KeychainError('Keychain returned empty token')
    return token


def mask_token(t):
    if not t:
        return '(empty)'
    return f'{t[:4]}... (len={len(t)})'


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--service')
    parser.add_argument('--account')
    parser.add_argument('--env-prefix')
    parser.add_argument('--check', action='store_true')
    parser.add_argument('--print-unsafe', action='store_true')
    args = parser.parse_args()

    try:
        service = args.service or _default_service()
        account = args.account or _default_account()
    except Exception as e:
        sys.stderr.write(f'[keychain_get_secret] config error: {e}\n')
        return 2
    try:
        t = get_secret(service=service, account=account)
    except KeychainError as e:
        sys.stderr.write(f'[keychain_get_secret] {e}\n')
        return e.exit_code

    if args.check:
        print(f'OK {mask_token(t)}')
    elif args.print_unsafe:
        sys.stdout.write(t)
    else:
        print(f'OK service={service} account={account} {mask_token(t)}')
        print('hint: use --print-unsafe to emit raw token (avoid in shared terminals)')
    return 0


if __name__ == '__main__':
    sys.exit(main())
